src/hooks/useComposedCssClasses.tsx

```diff
@@ -1,76 +1,54 @@
- * Replace: `{ icon: 'Blue' }`
- * Assign: `{ icon: 'Blue', button: 'Button' }`
+ * Suppose a component has built-in classes of `{ container: 'px-4 text-slate-700' }`.
  *
- * @public
- */
-export type CompositionMethod = 'merge' | 'replace' | 'assign';
-
-/**
- * A react hook which combines a component's built-in CSS classes with custom CSS classes.
+ * Passing in the custom classes:
+ *
+ * ```ts
+ * { container: 'text-red-200 mb-3' }
+ * ```
+ *
+ * results in the merged classes of:
+ *
+ * ```ts
+ * { container: 'px-4 text-red-200 mb-3' }
+ * ```
  *
  * @public
  *
- * @param builtInClasses - The component's built-in css classes
- * @param customClasses - The custom classes to combine with the built-in ones
- * @param cssCompositionMethod - The method of combining the built-in classes with the custom classes
+ * @param builtInClasses - The component's built-in tailwind classes
+ * @param customClasses - The custom tailwind classes to merge with the built-in ones
  * @returns The composed CSS classes
  */
-export function useComposedCssClasses<ClassInterface>(
+export function useComposedCssClasses<
+  ClassInterface extends Partial<Record<keyof ClassInterface & string, string>>
+>(
   builtInClasses: ClassInterface,
-  customClasses?: ClassInterface,
-  cssCompositionMethod?: CompositionMethod
-): ClassInterface | Theme {
+  customClasses?: Partial<ClassInterface>
+): ClassInterface {
   return useMemo(() => {
-    if (!isThemeObject(customClasses)) {
-      return builtInClasses;
+    const mergedCssClasses = { ...builtInClasses };
+    if (!customClasses) {
+      return mergedCssClasses;
     }
-    if (!isThemeObject(builtInClasses)) {
-      return customClasses ?? {};
-    }
-    const compose = getCssModulesCompose(cssCompositionMethod);
-    return composeTheme([{ theme: builtInClasses }, { theme: customClasses, compose }]);
-  }, [builtInClasses, customClasses, cssCompositionMethod]);
-}
-
-/**
- * Transforms the CompositionMethod types to the Compose types of the css-modules-theme library
- * @param cssCompositionMethod - The cssCompositionMethod method
- * @returns The css-modules-theme Compose type
- */
-function getCssModulesCompose(cssCompositionMethod: CompositionMethod = 'merge'): Compose {
-  return {
-    replace: Compose.Replace,
-    assign: Compose.Assign,
-    merge: Compose.Merge
-  }[cssCompositionMethod];
-}
-
-/**
- * Returns true if the object can be used as a css-modules-theme Theme
- * @param obj - The object to test
- * @returns Whether or not the object is a Theme object
- */
-function isThemeObject(obj: unknown): obj is Theme {
-  if (obj === null || typeof obj !== 'object') {
-    return false;
-  }
-  return true;
+    Object.keys(customClasses).forEach(key => {
+      const builtIn = builtInClasses[key];
+      const custom = customClasses[key];
+      if (!builtIn || !custom) {
+        mergedCssClasses[key] = builtIn ?? custom;
+      } else {
+        mergedCssClasses[key] = twMerge(builtIn, custom);
```

Comment:
I believe [twMerge ignores undefined/null values](https://github.com/dcastil/tailwind-merge#ignores-undefined-null-and-false-values) so I'm not sure if this check is needed?

Reply:
Oh so I wasn't sure if this was necessary when I wrote it so I looked at the source code to see if it might be worth it and they do call [mergeClasslist](https://github.com/dcastil/tailwind-merge/blob/main/src/lib/merge-classlist.ts) every time twMerge is called and also check and update a cache so I felt like eh may as well? But I can remove if you feel this is a premature optimization